csrc/aes_xts.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+#include "env.h"
+#include <openssl/evp.h>
+#include <jni.h>
+
+namespace AmazonCorrettoCryptoProvider {
+
+class AesXtsCipher {
+public:
+    AesXtsCipher(bool for_encryption, unsigned char const* key, unsigned char const* tweak)
+    {
+        ctx_ = EVP_CIPHER_CTX_new();
+        if (ctx_ == nullptr) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_CIPHER_CTX_new failed.");
+        }
+
+        if (for_encryption) {
+            if (EVP_EncryptInit_ex(ctx_, EVP_aes_256_xts(), nullptr, key, tweak) != 1) {
+                throw_openssl(EX_RUNTIME_CRYPTO, "EVP_EncryptInit_ex failed.");
+            }
+        } else {
+            if (EVP_DecryptInit_ex(ctx_, EVP_aes_256_xts(), nullptr, key, tweak) != 1) {
+                throw_openssl(EX_RUNTIME_CRYPTO, "EVP_DecryptInit_ex failed.");
+            }
+        }
+
+        // this method always returns 1
+        EVP_CIPHER_CTX_set_padding(ctx_, 0);
+    }
+
+    ~AesXtsCipher() { EVP_CIPHER_CTX_free(ctx_); }
+
+    void encrypt(unsigned char* input, int input_len, unsigned char* output)
+    {
+        int out_len = 0;
+        if (EVP_EncryptUpdate(ctx_, output, &out_len, input, input_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_EncryptUpdate failed.");
+        }
+
+        if (EVP_EncryptFinal_ex(ctx_, output + out_len, &out_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_EncryptFinal_ex failed.");
+        }
+    }
+
+    void decrypt(unsigned char* input, int input_len, unsigned char* output)
+    {
+        int out_len = 0;
+        if (EVP_DecryptUpdate(ctx_, output, &out_len, input, input_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_DecryptUpdate failed.");
+        }
+
+        if (EVP_DecryptFinal_ex(ctx_, output + out_len, &out_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_DecryptFinal_ex failed.");
+        }
+    }
```

Comment:
Should `encrypt`'s input and `decrypt`'s output pointers be const? The EVP functions being called use that convention in their own calling signatures, and you did in the constructor as well.

Reply:
The reason it's not const is that the same buffer for input can be used as output.

Reply:
Ack. Thanks for the clarification.